gips/datahandler/scheduler.py

```diff
@@ -73,62 +74,91 @@ def schedule_query ():
+        active = model.objects.filter(
             status__in=['scheduled', 'in-progress', 'retry']
         )
+        if driver is not None:
+            active = active.filter(driver=driver)
 
-        # check if jobs still alive
-        active_jobs = active.order_by('sched_id').distinct('sched_id')
+        active_jobs = active.values_list('sched_id', flat=True).distinct()
         for j in active_jobs:
-            if j.sched_id and queue.is_job_alive(j.sched_id):
-                Logger().log(
-                    'Previous {} fetch job still running. Moving on...'
-                    .format(driver)
-                )
-                return
-
-        # if anything leftover, clean it up
-        needs_cleanup = active.filter(
-            assetstatuschange__status='requested',
-        ).annotate(
-            num_retries=models.Count('assetstatuschange__status'),
-        )
-        if needs_cleanup.exists():
-            # something went wrong. not 'complete', but no running/scheduled job
-            Logger().log('Attempting to reschedule incomplete tasks')
-
-            retries = []
-            fails = []
-            for a in needs_cleanup:
-                if a.num_retries >= 3:
-                    fails.append(a)
-                    # if it has been retried 3 times, give up
-                    a.status = 'failed'
+            print "j is", j
+            if j and queue.is_job_alive(j):
+                is_something_running = True
+            else:
+                # find all tasks associated with this job
+                needs_cleanup = active.filter(sched_id=j)
+                if model == dbinv.models.Asset:
+                    needs_cleanup = needs_cleanup.filter(
+                        assetstatuschange__status='requested',
+                    ).annotate(
+                        num_retries=models.Count('assetstatuschange__status'),
+                    )
+                elif model == dbinv.models.Product:
+                    needs_cleanup = needs_cleanup.filter(
+                        productstatuschange__status='requested',
+                    ).annotate(
+                        num_retries=models.Count('productstatuschange__status'),
+                    )
                 else:
-                    retries.append(a)
-                    a.status = 'requested'
-                    a.sched_id = ''
-                a.save()
-
-            if fails:
-                Logger().log(
-                    "Fetch failed three times - give up\n{}"
-                    .format(pformat([(f.id, f.driver, f.asset, f.tile, f.date)
-                                     for f in fails]))
-                )
-            if retries:
-                Logger().log(
-                    "Retrying fetch\n{}"
-                    .format(pformat([(r.id, r.driver, r.asset, r.tile, r.date)
-                                     for r in retries]))
-                )
+                    raise # TODO raise something useful
```

Comment:
optional:  I like `RuntimeError('Unreachable line reached:  model was neither Product nor Asset')` similar to:

https://github.com/Applied-GeoSolutions/gips/blob/batch-scheduler/gips/datahandler/queue/queue.py#L43